tiledb/api/src/datatype/arrow.rs

```diff
@@ -46,169 +53,667 @@ impl ArrowPrimitiveTypeNative for f64 {
+            Ok(len) => match NonZeroU32::new(len) {
+                None => Res::None,
+                Some(nz) => Res::Exact(Datatype::Blob, CellValNum::Fixed(nz)),
+            },
+            Err(_) => Res::None,
+        },
+        ADT::FixedSizeList(ref item, ref len) => {
+            let len = match u32::try_from(*len).ok().and_then(NonZeroU32::new) {
+                Some(len) => len,
+                None => return Res::None,
+            };
+            if item.is_nullable() {
+                // no way to represent null values within a cell
+                Res::None
+            } else if item.data_type().primitive_width().is_none() {
+                /*
+                 * probably there are some cases we can handle,
+                 * but let's omit for now
+                 */
+                Res::None
+            } else if let Some(exact_datatype) = item
+                .metadata()
+                .get(ARROW_FIELD_METADATA_KEY_TILEDB_EXACT_MATCH)
+                .and_then(|s| Datatype::from_string(s))
+            {
+                Res::Exact(exact_datatype, CellValNum::Fixed(len))
+            } else {
+                match from_arrow(item.data_type()) {
+                    Res::None => Res::None,
+                    Res::Inexact(item, item_cell_val) => {
+                        let cell_val_num = match item_cell_val {
+                            CellValNum::Fixed(nz) => {
+                                match nz.checked_mul(len) {
+                                    None => return Res::None,
+                                    Some(nz) => CellValNum::Fixed(nz),
+                                }
+                            }
+                            CellValNum::Var => CellValNum::Var,
+                        };
+                        Res::Inexact(item, cell_val_num)
+                    }
+                    Res::Exact(item, item_cell_val) => {
+                        let cell_val_num = match item_cell_val {
+                            CellValNum::Fixed(nz) => {
+                                match nz.checked_mul(len) {
+                                    None => return Res::None,
+                                    Some(nz) => CellValNum::Fixed(nz),
+                                }
+                            }
+                            CellValNum::Var => CellValNum::Var,
+                        };
+                        Res::Exact(item, cell_val_num)
+                    }
+                }
+            }
+        }
+        ADT::LargeUtf8 => {
+            /*
+             * NB: arrow checks for valid UTF-8 but tiledb does not.
+             * This is not an exact conversion for that reason
+             * because we cannot guarantee invertibility.
+             */
+            Res::Inexact(Datatype::StringUtf8, CellValNum::Var)
+        }
+        ADT::LargeList(ref item) => {
+            if item.is_nullable() {
+                // no way to represent null values within a cell
```

Comment:
What? We've got the validity buffer just like they do?

Reply:
The validity buffer is for each cell.  For a var-size attribute the number of values per cell is variable but there is only one validity, for the whole cell.
The `item.is_nullable()` is about whether the values have a validity, not just the cell

Reply:
Ahhhhhh! Gotchya. Could you reword that comment to that effect as I totes blew right past that subtlety with the comment. Just something like "TileDB supports validity of the entire cell, but not for values within the cell so we can't support this". I'm guessing my brain read "TileDB has no way to represent validity" and then clicked off to "What? Yes it do!" before finishing the "within a cell" part.